### PR TITLE
DAOS-2660: object classes and types cleanup

### PR DESCRIPTION
**Pull request by gnailzenh**
- A bunch of code movement
  . move object types from daos_types.h to include/daos_obj.h
    and include/daos/object.h
  . move object APIs from daos_api.h to include/daos_obj.h
  . move property APIs and types to include/daos_prop.h
  . move a few object helper functions from src/common/misc.c
    to src/object/obj_utils.c
  . move daos_iod_t helper functions from common to object
  . move recx btree class from common to object

- Cleanup for exported object classes
  . formalize class ID
  . clearly separte EC and replication classes
  . add predefined class IDs

- Adjust reserved bits of object ID.
  . decrease object ID version from 16-bit to 4-bit
  . decrease class bits from 16 to 12
  . increase feature bits from 8 to 16
  . add ECHO and KV_FLAT as new feature bits

- Add a new parameter to daos_obj_generate_id(), which allows
  user to pass in another 32-bit integer as new paramter.
  This is reserved for support of user-defined redundancy group
  size/number in the future.

- Other cleanups
  . cleanup for struct daos_iom_t
  . remove DAOS_OF_DKEY/AKEY_HASHED, they are zeroes anyway
  . add a class ID converter in python, it can convert hard-coded
    old class ID to new class ID.

- TODO:
  . add class attributes for all predefined object classes
  . add classes to support of specified group size/number
  . remove obsolete object classes
  . use ECHO feature bit to replace echo object classes
  . move property API definition from common/misc.c to common/prop.c
  . move daos_obj_generate_id to obj_class.c, this also requires
    quite a bit of code cleanup because vos_test does not link to
    libdaos but it needs to generate ID.

Signed-off-by: Liang Zhen <liang.zhen@intel.com>

### CODE REVIEW

File: src/object/SConscript [hunk 1]
```diff
@@ -10,7 +10,8 @@ def scons():
     denv = env.Clone()
 
     # Common object code
-    common_tgts = denv.SharedObject(['obj_class.c', 'obj_rpc.c', 'obj_task.c'])
+    common_tgts = denv.SharedObject(['obj_class.c', 'obj_rpc.c', 'obj_task.c',
+				     'obj_utils.c'])
```
[the remark this answers is not in the document]
Comment (jolivier23) [reply src/object/SConscript]: Just add one space before 'obj_utils.c'.  It is supposed to line up with entries in the python array

File: src/include/daos_obj.h [hunk 2]
```diff
@@ -0,0 +1,786 @@
...
+	 * external arrays listed below. 1 for SV.
+	 */
+	unsigned int		 iom_nr;
+	/** Size of the single value or the record size */
+	daos_size_t		 iom_size;
+	/**
+	 * The recx with the lowest offset within the requested extents
+	 * daos_io_desc::iod_recxs
+	 */
+	struct daos_recx	 iom_recx_lo;
```
Comment (jolivier23) [inline src/include/daos_obj.h]: why are these needed?   if iom_recxs is simply sorted, can't the user just infer the lo and the hi ?

File: src/include/daos_obj.h [hunk 3]
```diff
@@ -0,0 +1,786 @@
+/**
```
Comment (jolivier23) [inline src/include/daos_obj.h]: It would probably be easier to review if we did the reorganization in a separate patch and then added the new object classes.   Probably not worth it at this point but would make changes like this simpler in the future.

File: src/vos/tests/vos_tests.c [hunk 4]
```diff
@@ -55,16 +55,36 @@ print_usage()
...
 	int	i;
+	int	j;
 
 	failed += run_pool_test();
 	failed += run_co_test();
-	for (i = 0; i != DAOS_OF_MASK; i++)
-		failed += run_io_test(i, keys, nest_iterators);
+	for (i = 0; dkey_feats[i] >= 0; i++) {
+		for (j = 0; akey_feats[j] >= 0; j++) {
+			feats = dkey_feats[i] | akey_feats[j];
```
Comment (jolivier23) [inline src/vos/tests/vos_tests.c]: nice cleanup

File: src/include/daos_obj.h (hunk 2 above)
Comment (mchaarawi) [reply src/include/daos_obj.h]: we don't want to fetch all recxs in some situations, and just the hi / lo

File: src/client/api/object.c [hunk 5]
```diff
@@ -165,7 +165,7 @@ daos_obj_query_key(daos_handle_t oh, daos_handle_t th, uint32_t flags,
 int
 daos_obj_fetch(daos_handle_t oh, daos_handle_t th, daos_key_t *dkey,
 	       unsigned int nr, daos_iod_t *iods, d_sg_list_t *sgls,
-	       daos_iom_t *maps, daos_event_t *ev)
+	       struct daos_io_map *maps, daos_event_t *ev)
```
Comment (mchaarawi) [inline src/client/api/object.c]: i think we should use the typedef'd daos_iom_t.
otherwise the API looks weird with daos_iod_t, and then struct daos_io_map

File: src/include/daos_obj.h (hunk 2 above)
Comment (gnailzenh) [reply src/include/daos_obj.h]: indeed, require user to always allocate array is not very nice, specially when most of them don't really care about those recxs

File: src/include/daos_obj.h (hunk 3 above)
Comment (gnailzenh) [reply src/include/daos_obj.h]: you are right, moving things around can be in a separate patch, I underestimated the work, I thought just a bunch of renames would be sufficient, but it took me way more time eventually.

File: src/vos/tests/vos_tests.c (hunk 4 above)
Comment (gnailzenh) [reply src/vos/tests/vos_tests.c]: since I'm adding more feature bits, so vos_test took forever because it repeats thousands of times.

File: src/client/api/object.c (hunk 5 above)
Comment (gnailzenh) [reply src/client/api/object.c]: our plan is removing most of typedefs gradually (we will keep typedef for small structures like object ID, d_iov_t, but remove typedefs for data structures which are large and not suitable for putting in function call stack).

So I'm keeping both structure name and typedef name for some structures, we can use structure name in new code, and removing typedefed name incrementally, instead of generating another a few thousands of lines patch which renames everything.

**Comment by gnailzenh** [comment]
I have to apply a python patch to make ftest work

**Comment by gnailzenh** [comment]
I just reverted those typedef changes, let's keep them for now, and decide later.

**Comment by mchaarawi** [comment]
> Mohamad, if you think it's a bad time to do this, I will revert the structure name from this patch, and we can discuss about it again.

yea let's postpone these changes. I'd rather we make all the change in the API at once rather than multiple passes.

File: src/include/daos_obj_class.h [hunk 6]
```diff
@@ -0,0 +1,525 @@
...
+	/**
+	 * Replicated object with maximum number of redundancy groups.
+	 * Number of redundancy groups of the class is chosen by DAOS
+	 * based on the current size of the pool.
+	 */
+	OC_RP_MAX,
+
+	/**
+	 * Object classes protected by replication which supports Scalable
+	 * Fetch (SF)
```
Comment (liuxuezhao) [inline src/include/daos_obj_class.h]: for _SF_ obj classes, how to determine the number of replicas? (domain_nr?)

File: src/include/daos_obj_class.h [hunk 7]
```diff
@@ -0,0 +1,525 @@
...
+	 * It has many replicas so it is very slow for update.
+	 */
+	OC_RP_XSF	= 80,
+
+	/**
+	 * Object classes protected by erasure code
+	 */
+	/**
+	 * Tiny object protected by EC
+	 * This object class has one redundancy group
```
Comment (liuxuezhao) [inline src/include/daos_obj_class.h]: how to determine the number of original k/p, and len?

File: src/include/daos_obj_class.h [hunk 8]
```diff
@@ -0,0 +1,525 @@
...
+	 * Number of redundancy groups of the class is chosen by DAOS
+	 * based on the current size of the pool.
+	 */
+	OC_EC_LARGE,
+	/**
+	 * EC object with maximum number of redundancy groups.
+	 * Number of redundancy groups of the class is chosen by DAOS
+	 * based on the current size of the pool.
+	 */
+	OC_EC_MAX,
```
Comment (liuxuezhao) [inline src/include/daos_obj_class.h]: maybe can add "D_CASSERT(OC_EC_MAX < OC_S1 && OC_SX < OC_RP_2G1 ...)" at somewhere.

**Comment by gnailzenh** [comment]
Hmm, I don't know what's going on but it shows "rebuild test failed to start...", I have to push it again, sorry for the noise
